### PR TITLE
Fix problematic reflection usage

### LINKED ISSUE #18782: com.hazelcast.replicatedmap.ReplicatedMapReorderedReplicationTest.testNonConvergingReplicatedMaps
_4.2.z_ (commit e22aab3cffbb36863a91276dd2506b4fcf4a7f6c)

Failed on Zulu JDK 16: http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-4.maintenance-ZuluJDK16-nightly/2/testReport/com.hazelcast.replicatedmap/ReplicatedMapReorderedReplicationTest/testNonConvergingReplicatedMaps/

Stacktrace:
```
java.lang.NoSuchFieldException: modifiers
	at java.base/java.lang.Class.getDeclaredField(Class.java:2549)
	at com.hazelcast.replicatedmap.ReplicatedMapReorderedReplicationTest.updateFactory(ReplicatedMapReorderedReplicationTest.java:157)
	at com.hazelcast.replicatedmap.ReplicatedMapReorderedReplicationTest.testNonConvergingReplicatedMaps(ReplicatedMapReorderedReplicationTest.java:78)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:831)
```

Standard output:
```
BuildInfo right after testNonConvergingReplicatedMaps(com.hazelcast.replicatedmap.ReplicatedMapReorderedReplicationTest): BuildInfo{version='4.2.1-SNAPSHOT', build='20210526', buildNumber=20210526, revision=e22aab3, enterprise=false, serializationVersion=1}
Hiccups measured while running test 'testNonConvergingReplicatedMaps(com.hazelcast.replicatedmap.ReplicatedMapReorderedReplicationTest):'
16:43:45, accumulated pauses: 827 ms, max pause: 582 ms, pauses over 1000 ms: 0
```

### PR DESCRIPTION
**Pull request by ramizdundar**
There is a nightly tests failing in Zulu JDK 16 #18782. This is because after JDK 12 one can't access fields of `Field` class. We want to do this in order to change access modifiers of `ReplicatedMapDataSerializerHook#FACTORY`. This PR modifies the code so that `Unsafe` is used in order to change modifiers of `ReplicatedMapDataSerializerHook#FACTORY`

Fixes #18782

This could be backported

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Request reviewers if possible
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases

### CODE REVIEW

**Comment by ramizdundar** [comment]
Related test passed: http://jenkins.hazelcast.com/view/Hazelcast/job/Hazelcast-custom-branch-ZuluJDK16-nightly/3/

**Review by jbartok — APPROVED** [review]

File: hazelcast/src/test/java/com/hazelcast/replicatedmap/ReplicatedMapReorderedReplicationTest.java [hunk 1]
```diff
@@ -193,4 +189,16 @@ public void run() throws Exception {
...
+    private void updateFactoryField(DataSerializableFactory factory) throws Exception {
+        // Get Unsafe object
+        final Field unsafeField = Unsafe.class.getDeclaredField("theUnsafe");
+        unsafeField.setAccessible(true);
+        final Unsafe unsafe = (Unsafe) unsafeField.get(null);
+
+        // Update the field
+        final Object staticFieldBase = unsafe.staticFieldBase(field);
+        final long staticFieldOffset = unsafe.staticFieldOffset(field);
+        unsafe.putObject(staticFieldBase, staticFieldOffset, factory);
```
Comment (ufukyilmaz) [inline hazelcast/src/test/java/com/hazelcast/replicatedmap/ReplicatedMapReorderedReplicationTest.java:202]:
```suggestion
        // Update the field
        final Object staticFieldBase = UnsafeUtil.UNSAFE.staticFieldBase(field);
        final long staticFieldOffset = UnsafeUtil.UNSAFE.staticFieldOffset(field);
        UnsafeUtil.UNSAFE.putObject(staticFieldBase, staticFieldOffset, factory);
```
we can use unsafe util here. Also, could you handle the imports?

**Comment by ramizdundar** [comment]
http://jenkins.hazelcast.com/view/Hazelcast/job/Hazelcast-custom-branch-ZuluJDK16-nightly/4/

**Review by ufukyilmaz — APPROVED** [review]
It looks good to me.

**Review by ahmetmircik — APPROVED** [review]
LGTM

**Pull request merged**